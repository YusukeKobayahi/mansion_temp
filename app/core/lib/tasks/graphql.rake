# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
