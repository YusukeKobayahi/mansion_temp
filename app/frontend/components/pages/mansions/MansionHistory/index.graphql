fragment MsLibraryMansionsMansionHistory on Mansion {
  uniqueCode
  name
  constructedIn
  age
  prefecture {
    name
    jisCode
  }
  city {
    name
    jisCode
  }
  town {
    name
  }
  street {
    name
  }
  blockNumber
  salesHistorySummaries {
    layout
    minPrice
    maxPrice
  }
  access {
    id
    stationId
    stationLineName
    stationName
    stationWalkingMinutes
    lineId
  }
}
