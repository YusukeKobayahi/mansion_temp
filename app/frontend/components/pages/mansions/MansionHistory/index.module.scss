@import "~/styles/variables";
@import "~/styles/utils";

.container {
  margin-top: 50px;
}
