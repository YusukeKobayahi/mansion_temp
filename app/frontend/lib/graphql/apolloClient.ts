// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
