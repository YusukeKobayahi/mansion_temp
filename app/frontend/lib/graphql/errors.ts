export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
