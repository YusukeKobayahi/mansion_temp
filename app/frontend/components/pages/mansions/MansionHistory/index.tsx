import { useEffect, useState } from "react";
import { useDeepCompareMemo } from "use-deep-compare";
import compact from "lodash/compact";
import styles from "./index.module.scss";
import SalesSummaryBuilder, {
  LayoutSummary,
} from "~/lib/mansions/SalesSummaryBuilder";
import {
  buildLibraryPagePath,
  buildAccess,
  buildConstructedDate,
} from "~/lib/utils";
import { MsLibraryMansionsMansionHistoryFragment } from "~/graphql/generated";
import MansionSlider from "~/components/pages/mansions/MansionSlider";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import { GoogleAnalyticsEvent } from "~/lib/types";

export interface MansionHistoryProps {
  data: MsLibraryMansionsMansionHistoryFragment;
}

type MansionDataProps = {
  href?: string;
  name?: string;
  address?: string;
  salesInfo?: LayoutSummary[];
  constructedDate?: string;
  access?: string;
};

const MansionHistory: React.FC<MansionHistoryProps> = ({
  data,
}: MansionHistoryProps) => {
  const mansion = data;
  const href = buildLibraryPagePath("mansions", mansion.uniqueCode);
  const name = mansion.name;
  const builder = new SalesSummaryBuilder(mansion.salesHistorySummaries);
  const salesInfo = builder.buildLayoutSummaries();
  const access = buildAccess(mansion.access);
  const constructedDate = buildConstructedDate(
    mansion.constructedIn,
    mansion.age
  );
  const address = compact([
    mansion.prefecture.name,
    mansion.city.name,
    mansion.town.name,
    mansion.street?.name,
    mansion.blockNumber,
  ]).join("");

  const mansionData: MansionDataProps = {
    href,
    name,
    address,
    salesInfo,
    constructedDate,
    access,
  };
  const maxStrageNum = 10;
  const updateMansionHistory = () => {
    try {
      let mansionHistory: MansionDataProps[] = JSON.parse(
        localStorage.getItem("mansionHistory")
          ? String(localStorage.getItem("mansionHistory"))
          : "[]"
      );
      mansionHistory = mansionHistory.filter(
        (el) => el.href !== mansionData.href
      );
      mansionHistory.push(mansionData);
      if (mansionHistory.length > maxStrageNum)
        mansionHistory = mansionHistory.slice(0, maxStrageNum);
      localStorage.setItem("mansionHistory", JSON.stringify(mansionHistory));
    } catch {
      // ESLintでエラーが出てしまうため、コメントアウト
    }
  };
  const [mansionHistory, setMansionHistory] = useState<MansionDataProps[]>([]);
  useEffect(() => {
    updateMansionHistory();
    let mansionHistory;
    try {
      mansionHistory = JSON.parse(
        localStorage.getItem("mansionHistory")
          ? String(localStorage.getItem("mansionHistory"))
          : "[]"
      );
    } catch {
      mansionHistory = [];
    }
    setMansionHistory(mansionHistory);
  }, [mansion.uniqueCode]);

  const displayMansionHistory = [...mansionHistory].reverse().splice(1);
  const eventParams: GoogleAnalyticsEvent = {
    action: "msl_link_click",
    category: "msl_shosai",
    label: "msl_history_bukken",
    value: 1,
  };
  const mansionCards = useDeepCompareMemo(() => {
    return displayMansionHistory.map(
      ({ href, name, constructedDate, salesInfo, address, access }) => {
        return {
          href,
          name,
          constructedDate,
          salesInfo,
          address,
          access,
          eventParams,
        };
      }
    );
  }, [mansionHistory]);

  return mansionCards.length > 0 ? (
    <section className={styles.container}>
      <BoxHeading>確認した物件</BoxHeading>
      <MansionSlider mansionCards={mansionCards} />
    </section>
  ) : null;
};

export default MansionHistory;
