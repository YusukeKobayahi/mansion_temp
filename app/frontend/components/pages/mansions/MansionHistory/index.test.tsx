import { render } from "@testing-library/react";
import mockData from "~/components/pages/mansions/fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import MansionHistory from "~/components/pages/mansions/MansionHistory";

describe("components/pages/mansions/MansionHistory", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<MansionHistory data={mock.mansion} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
