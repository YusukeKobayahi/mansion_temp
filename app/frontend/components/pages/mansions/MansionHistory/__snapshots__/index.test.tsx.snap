// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/MansionHistory 正常に表示される 1`] = `<DocumentFragment />`;
